extensions/cornerstone/src/tools/ImageOverlayViewerTool.tsx

```diff
@@ -0,0 +1,307 @@
+      bitIdx = 0;
+      byteIdx++;
+    } else {
+      bitIdx++;
+    }
+  }
+  ctx.putImageData(imageData, 0, 0);
+
+  return canvas.toDataURL();
+};
+
+/**
+ *
+ * @param imageId
+ * @param overlayMetadata
+ * @param color {number[]} - color (r,g,b,a), default color: gray
+ * @returns
+ */
+const getCachedStat = async (
+  imageId: string,
+  overlayMetadata: any[],
+  color: number[] = [127, 127, 127, 255]
+): Promise<CachedStat> => {
+  if (
+    !cachedStats[imageId] ||
+    !isSameColor(cachedStats[imageId].color, color)
+  ) {
+    const overlays = await Promise.all(
+      overlayMetadata
+        .filter(overlay => overlay.pixelData)
+        .map(async (overlay, idx) => {
+          let pixelData = null;
+          if (overlay.pixelData.Value) {
+            pixelData = overlay.pixelData.Value;
+          } else if (overlay.pixelData.retrieveBulkData) {
+            pixelData = await overlay.pixelData.retrieveBulkData();
+          }
+
+          if (!pixelData) return;
+
+          const dataUrl = renderOverlayToDataUrl(
+            { width: overlay.columns, height: overlay.rows },
+            color,
+            pixelData
+          );
+
+          return {
+            ...overlay,
+            _id: guid(),
+            dataUrl, // this will be a data url expression of the rendered image
+            color,
+          };
+        })
+    );
+
+    cachedStats[imageId] = {
+      color: color,
+      overlays: overlays.filter(overlay => overlay),
+    };
+  }
+
+  return cachedStats[imageId];
+};
+
+/**
+ * Image Overlay Viewer tool is not a traditional tool that requires user interactin.
+ * But it is used to display Pixel Overlays. And it will provide toggling capability.
+ *
+ * The documentation for Overlay Plane Module of DICOM can be found in [C.9.2 of
+ * Part-3 of DICOM standard](https://dicom.nema.org/medical/dicom/2018b/output/chtml/part03/sect_C.9.2.html)
+ *
+ * Image Overlay rendered by this tool can be toggled on and off using
+ * toolGroup.setToolEnabled() and toolGroup.setToolDisabled()
+ */
+class ImageOverlayViewerTool extends BaseTool {
+  static toolName = 'ImageOverlayViewer';
+
+  _mode = 'Disabled';
+  _renderingViewport: any;
+
+  constructor(
+    toolProps = {},
+    defaultToolProps = {
+      supportedInteractionTypes: [],
+      configuration: {
+        fillColor: [255, 127, 127, 255],
+      },
+    }
+  ) {
+    super(toolProps, defaultToolProps);
+  }
+
+  onSetToolEnabled = (): void => {
+    this._mode = 'Enabled';
+  };
+
+  onSetToolDisabled = (): void => {
+    this._mode = 'Disabled';
+  };
+
+  renderAnnotation = (enabledElement, svgDrawingHelper) => {
+    // overlays are toggled off by configuration
+    if (this._mode !== 'Enabled') return false;
+
+    const { viewport } = enabledElement;
+    this._renderingViewport = viewport;
+
+    const imageId = this._getReferencedImageId(viewport);
+    if (!imageId) return;
+
+    const { overlays } = metaData.get('overlayPlaneModule', imageId);
```

Comment:
we should cache this, as we hit this A LOT! so `this.overlayCache.get(imageId)` if not found grab from metadata and add it to the cache